stm32h5 run application in external flash memory XIP

FRASTM opened the pull request:
the stm32h5 disco kit board can run an application in external flash

Requires the https://github.com/zephyrproject-rtos/zephyr/pull/88646 which removes the warning in the DTC

With this PR it enable the memorymapped mode on the external flash memory so that XIP is possible on the external NOR octoflash

Two samples are running in XIP on the external flash

samples/application_development/code_relocation_nocopy/
samples/boards/st/hello_world_xip/ built with mcu boot https://github.com/mcu-tools/mcuboot/pull/2323


Fixes https://github.com/zephyrproject-rtos/zephyr/issues/87531

This PR completes the https://github.com/zephyrproject-rtos/zephyr/pull/88051  for the stm32h5 serie

FRASTM committed:
drivers: clock: stm32H5 clock control of the PLL in MemoryMapped mode

Do not disable the PLL clock if it sources the XSPI and if the external
flash is executing in Place. After mcuboot reset, the code is executed
on the external flash, through the xspi.
The CONFIG_STM32_APP_IN_EXT_FLASH is set and will avoid re-config
of the PLL which is sourcing the XSPI peripheral. When eXecuting in Place
on this external NOR, it must not disable its own clock source (PLL).

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
west build -b stm32h573i_dk samples/application_development/code_relocation_nocopy/
west flash  : stm32cubeprogrammer runner is taking the external loader to download into the external NOR (0x90000000)
```
*** Booting Zephyr OS build v4.1.0-2350-g86df576c4134 ***
Address of main function 0x80005e1
Address of function_in_ext_flash 0x90000001
Address of var_ext_sram_data 0x200000a0 (10)
Address of function_in_sram 0x20000001
Address of var_sram_data 0x200000a4 (10)
Hello World! stm32h573i_dk
```

FRASTM commented:
west build -p -b stm32h573i_dk samples/boards/st/hello_world_xip/ --sysbuild --     -DSB_CONFIG_BOOTLOADER_MCUBOOT=y

requires https://github.com/mcu-tools/mcuboot/pull/2154
download with stm32cubeprogrammer GUI
the build/mcuboot/zephyr/zephyr.bin at internal flash address 0x8000000
the build/hello_world_xip/zephyr/zephyr.signed.bin at external flash address 0x90000000

FRASTM commented:
Define larger partition in the external NOR flash (consequently adjusting the SECTOR size for the mcu boot)

FRASTM commented:
CI failure "invalid length for memory region EXTFLASH" given by the samples/application_development/code_relocation_nocopy/linker_arm_nocopy.ld

tpambor commented:
See https://github.com/zephyrproject-rtos/zephyr/pull/88392 for a fix.

FRASTM commented:
set the DNM mabel until the https://github.com/zephyrproject-rtos/zephyr/pull/88646 is merged

FRASTM commented:
Adding the commit to retrieve the CONFIG_FLASH_BASE_ADDRESS from the DTS by the soc/st/stm32/Kconfig.defconfig

FRASTM commented:
set the DNM mabel until the https://github.com/zephyrproject-rtos/zephyr/pull/88051 is merged

FRASTM commented:
Adding the configuration to run 
`west build -p -b stm32h573i_dk samples/sysbuild/with_mcuboot/ --sysbuild`
on the target board.
Use the stm32CubeProgrammer to download 

1. build/mcuboot/zephyr/zephyr.bin in internal flash at 0x8000000
2. build/with_mcuboot/zephyr/zephyr.signed.bin in external flash at 0x90000000


```
*** Booting MCUboot v2.1.0-rc1-277-g5ba3b7962ee8 ***
*** Using Zephyr OS build v4.1.0-3467-g1fee1c88b558 ***
*** Booting Zephyr OS build v4.1.0-3467-g1fee1c88b558 ***
Address of sample 0x90000000
Hello sysbuild with mcuboot! stm32h573i_dk

```

FRASTM committed:
boards: arm: stm32h573 disco kit move partition to ext flash

Add the XSPI 2 which is an octoSPI connection to a octo NOR flash
mx25lm51245 on the stm32h573i_dk disco kit.
Define the partition on the 64MBytes space of the external flash.
Use the STM32Cube programmer to flash with the corresponding
external loader for XiP mode.
Removing usb_device will avoid test feature for that board.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on boards/st/stm32h573i_dk/stm32h573i_dk.dts:
```diff
@@ -22,9 +22,9 @@
 		zephyr,shell-uart = &usart1;
 		zephyr,sram = &sram1;
 		zephyr,flash = &flash0;
-		zephyr,code-partition = &slot0_partition;
```
keeping the line

FRASTM commented:
> Firstly. Thanks a lot for this fix. But when trying to enable CONFIG_FLASH in the hello_world_xip sample there's an xspi init error. Maybe because the flash is being reinitialized in the application? I think there should be a way to disable the flash initialisation when the using mcuboot, because to update the images CONFIG_FLASH needs to be enabled for the img_manager.

I tested again the hello_world_xip sample
```
*** Booting MCUboot v2.1.0-rc1-315-ge3b8bd449b18 ***
*** Using Zephyr OS build v4.1.0-5079-g974edcd9c7ce ***
I: Starting bootloader
I: Primary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Secondary image: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: none
I: Image index: 0, Swap type: none
I: Bootloader chainload address offset: 0x0
I: Image version: v0.0.0
I: Jumping to the first image slot
*** Booting Zephyr OS build v4.1.0-5079-g974edcd9c7ce ***
Hello World! from external flash  stm32h573i_dk
--> PC at 0x9000092a
```

In the drivers/clock_control/clock_stm32_ll_h5.c, there is a sequence to avoid disabling the pll of the XSPI instance when this XSPI instance is in Memory Mapped mode (else you disable the clock of the peripheral you are executing)
And in the flash_stm32_xspi driver, there is a sequence to pass the flash_stm32_xspi_init when the xspi instance is executing in Memory Mapped mode.

FRASTM commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -434,6 +434,21 @@ static int set_up_plls(void)
+	 * that runs in memmap mode when (Q/O)SPI uses PLL
+	 * as its clock source.
+	 */
+#if defined(XSPI1) || defined(XSPI2)
```
removed

FRASTM commented:
with comments addressed

FRASTM commented:
> Now I get this Error during compiling:
> 
> ```
> C:/Users/jschaller/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h5.c:444:13: warning: implicit declaration of function 'LL_RCC_GetOSPIClockSource'; did you mean 'LL_RCC_GetSPIClockSource'? [-Wimplicit-function-declaration]
>   444 |         if (LL_RCC_GetOSPIClockSource(LL_RCC_OSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL1Q) {
>       |             ^~~~~~~~~~~~~~~~~~~~~~~~~
>       |             LL_RCC_GetSPIClockSource
> C:/Users/jschaller/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h5.c:444:39: error: 'LL_RCC_OSPI_CLKSOURCE' undeclared (first use in this function); did you mean 'LL_RCC_SPI6_CLKSOURCE'?
>   444 |         if (LL_RCC_GetOSPIClockSource(LL_RCC_OSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL1Q) {
>       |                                       ^~~~~~~~~~~~~~~~~~~~~
>       |                                       LL_RCC_SPI6_CLKSOURCE```
> ```

hum, this comes after https://github.com/zephyrproject-rtos/zephyr/pull/88579#discussion_r2132115169

FRASTM commented:
> > > Now I get this Error during compiling:
> > > ```
> > > C:/Users/jschaller/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h5.c:444:13: warning: implicit declaration of function 'LL_RCC_GetOSPIClockSource'; did you mean 'LL_RCC_GetSPIClockSource'? [-Wimplicit-function-declaration]
> > >   444 |         if (LL_RCC_GetOSPIClockSource(LL_RCC_OSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL1Q) {
> > >       |             ^~~~~~~~~~~~~~~~~~~~~~~~~
> > >       |             LL_RCC_GetSPIClockSource
> > > C:/Users/jschaller/zephyrproject/zephyr/drivers/clock_control/clock_stm32_ll_h5.c:444:39: error: 'LL_RCC_OSPI_CLKSOURCE' undeclared (first use in this function); did you mean 'LL_RCC_SPI6_CLKSOURCE'?
> > >   444 |         if (LL_RCC_GetOSPIClockSource(LL_RCC_OSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL1Q) {
> > >       |                                       ^~~~~~~~~~~~~~~~~~~~~
> > >       |                                       LL_RCC_SPI6_CLKSOURCE```
> > > ```
> > 
> > 
> > hum, this comes after [#88579 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/88579#discussion_r2132115169) I'm going to test something else...
> 
> What do you mean with that. Is the problem on my side or in the current state of this PR?

It has to be changed with 
`	if (LL_RCC_GetOCTOSPIClockSource(LL_RCC_OCTOSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL1Q) {
`and 
`	if (LL_RCC_GetOCTOSPIClockSource(LL_RCC_OCTOSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL2R) {
`

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -434,17 +434,23 @@ static int set_up_plls(void)
 
 #if defined(STM32_PLL_ENABLED)
+
+#if defined(CONFIG_STM32_MEMMAP) && defined(CONFIG_BOOTLOADER_MCUBOOT)
```
Intentionally removing switching sysclk source to HSI?
If so, I think it should be mentioned in the commit message.
Also if so, can remove `clock_switch_to_hsi()` local helper function.

FRASTM replied:
switching the clock to HSI in case the pll1 is not the source clock of the xspi peripheral
--> `#else` part

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -434,17 +434,23 @@ static int set_up_plls(void)
-
+	if (LL_RCC_GetOCTOSPIClockSource(LL_RCC_OCTOSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL2R) {
+		LL_RCC_PLL2_Disable();
+	}
```
The part related to PLL2 should go in `#if defined(STM32_PLL2_ENABLED)`/`#endif` condition instructions 50 lines below.

FRASTM committed:
samples: sysbuild: with mcu_boot on the stm32h573i_dk

Run the sample to execute in place on the external flash of the
stm32h573 disco kit.
with_boot is build and linked for the external flash.
Download the mcuboot zephyr.bin in internal memory.
Download the with_mcuboot zephyr.signed.bin in external memory.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

etienne-lms commented on boards/st/stm32h573i_dk/stm32h573i_dk.dts:
```diff
@@ -406,9 +392,17 @@
+				label = "image-0";
+				reg = <0x00000000 DT_SIZE_M(16)>;
+			};
+			slot1_partition: partition@1000000 {
```
add an empty line between subnodes.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -506,6 +517,16 @@ static int set_up_plls(void)
 #if defined(STM32_PLL2_ENABLED)
+#if defined(CONFIG_STM32_MEMMAP) && defined(CONFIG_BOOTLOADER_MCUBOOT)
+	/*
+	 * If MEMMAP and MCUBOOT, then the OCTOSPI is necessarily also defined, 
```
Compliance complains about the trailing space char at line end.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -434,6 +434,17 @@ static int set_up_plls(void)
+	 * when (Q/O)SPI uses PLL as its clock source but exit.
+	 */
+	if (LL_RCC_GetOSPIClockSource(LL_RCC_OSPI_CLKSOURCE) == LL_RCC_OSPI_CLKSOURCE_PLL1Q) {
+		return 0;
```
I think returning now is not awaited. Maybe PLL2 and PLL3 are also to be enabled.
Implementation would maybe be simpler with 3 local helper functions: `set_up_pll1()`, `set_up_pll2()` and `set_up_pll3()`.

Jowllel commented:
Since the latest commits my code gets stuck at:
```
I: Bootloader chainload address offset: 0x0
I: Image version: v1.0.0
I: Jumping to the first image slot
```

This also happens when I try to run the sysbuild with_mcuboot sample on the stm32h573i_dk board.

Additionally, I think the lines in the set_up_plls functions that detect if ospi uses the respective plls should look like this:
```
if (LL_RCC_GetOCTOSPIClockSource(LL_RCC_OCTOSPI_CLKSOURCE) == LL_RCC_OSPI_CLKSOURCE_PLL1Q)
```
instead of OSPIClockSource (These symbols don't exist)

FRASTM commented:
> Since the latest commits my code gets stuck at:
> 
> ```
> I: Bootloader chainload address offset: 0x0
> I: Image version: v1.0.0
> I: Jumping to the first image slot
> ```
> 
> This also happens when I try to run the sysbuild with_mcuboot sample on the stm32h573i_dk board.
> 
> Additionally, I think the lines in the set_up_plls functions that detect if ospi uses the respective plls should look like this:
> 
> ```
> if (LL_RCC_GetOCTOSPIClockSource(LL_RCC_OCTOSPI_CLKSOURCE) == LL_RCC_OSPI_CLKSOURCE_PLL1Q)
> ```
> 
> instead of OSPIClockSource (These symbols don't exist)

Yes, I was preparing the branch for that and also matching with the https://github.com/mcu-tools/mcuboot/pull/2323
Fixing this now
Then I will prepare a new version including https://github.com/zephyrproject-rtos/zephyr/pull/88579#discussion_r2166136184

FRASTM commented:
next step : (end of july) : going to adapt the pll clock for the stm32H5 and match with the https://github.com/mcu-tools/mcuboot/pull/2323

FRASTM commented on samples/boards/st/hello_world_xip/README.rst:
```diff
@@ -0,0 +1,65 @@
+   :zephyr-app: samples/boards/st/hello_world_xip
+   :host-os: unix
+   :board: stm32h573i_dk
+   :goals: run
```
done

FRASTM commented on samples/boards/st/hello_world_xip/src/main.c:
```diff
@@ -0,0 +1,58 @@
+	int ret;
+
+	printk("Hello World! from external flash  %s\n", CONFIG_BOARD);
+	printf("--> PC at 0x%x\n", __get_PC());
```
using "printf()"

FRASTM commented on samples/boards/st/hello_world_xip/README.rst:
```diff
@@ -0,0 +1,65 @@
+
+.. code-block:: console
+
+    Hello World! from external flash b_u585i_iot02a
```
removed

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -435,17 +435,21 @@ static int set_up_plls(void)
-	if (LL_RCC_GetSysClkSource() == LL_RCC_SYS_CLKSOURCE_STATUS_PLL1) {
-		clock_switch_to_hsi();
+	if (LL_RCC_GetOCTOSPIClockSource(LL_RCC_OCTOSPI_CLKSOURCE) != LL_RCC_OSPI_CLKSOURCE_PLL1Q) {
+		LL_RCC_PLL1_Disable();
```
We still need to switch to HSI IMHO.

See how this is handled in **clock_stm32_ll_h7.c**.

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -501,12 +504,27 @@ static int set_up_plls(void)
+	goto setup_pll2;
 #endif /* STM32_PLL_ENABLED */
-
+setup_pll2:
```
I meant:
```c
	(...)
#endif /* STM32_PLL_ENABLED */

setup_pll2:
	(...)
```

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -554,11 +576,15 @@ static int set_up_plls(void)
-#endif /* STM32_PLL2_ENABLED */
 
+	goto setup_pll3;
+#endif /* STM32_PLL2_ENABLED */
+setup_pll3:
```
ditto here;
```suggestion
#endif /* STM32_PLL2_ENABLED */

setup_pll3:

FRASTM replied:
new line added

FRASTM commented:
I see the condition to "not disable the PLL if the OCTOSPI is clocked by the PLL" is wrong

1. when building the mcuboot, CONFIG_STM32_MEMMAP is set because the external octo flash is configured in Memory mapped mode, but CONFIG_BOOTLOADER_MCUBOOT is not set : in this case, the pll setup must be done at init
2. when building the application, CONFIG_STM32_MEMMAP is not set because the code is being executed from the external flash, the external octo flash is already configured in Memory mapped mode and must not be changed, but CONFIG_BOOTLOADER_MCUBOOT is set : in this case the pll which sources the octospi clock must not be changed

FRASTM commented:
> > I see the condition to "not disable the PLL if the OCTOSPI is clocked by the PLL" (CONFIG_STM32_MEMMAP && CONFIG_BOOTLOADER_MCUBOOT) was wrong
> > 
> > 1. when building the mcuboot, CONFIG_STM32_MEMMAP is set because the external octo flash is being configured in Memory Mapped mode, but CONFIG_BOOTLOADER_MCUBOOT is not set : in this case, the pll setup must be done at init
> > 2. when building the application, CONFIG_STM32_MEMMAP is not set because the code is being executed from the external flash, the external octo flash is _already_ configured in Memory Mapped mode and must not be changed. The octospi is not supposed to be re-initialized, its source clock must not be changed, but CONFIG_BOOTLOADER_MCUBOOT is set : in this case the pll which sources the octospi clock must not be altered
> 
> To detect that the Zephyr application being built is MCUboot, use CONFIG_MCUBOOT
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/4852719f98b3bac9c05196d9855d897f47e3c82f/modules/Kconfig.mcuboot#L11
> 
> > Ideally, we should find in zephyr a CONFIG flag to detect that it is running on the external flash after jumping from mcuboot. Could we consider this is CONFIG_BOOTLOADER_MCUBOOT
> 
> That would be `CONFIG_BOOTLOADER_MCUBOOT && CONFIG_STM32_MEMMAP && CONFIG_XIP`

Not directly: the CONFIG_STM32_MEMMAP is for configuring the external xspi flash.Of course it is set when building the mcuboot. When building a zephyr application like samples/sysbuild/with_mcuboot or  samples/boards/st/hello_world_xip, there is no reason to have the xspi node enabled, there is no reason to have the CONFIG_STM32_MEMMAP set (see "depends on").

Jowllel commented:
I just tested the hello_world_xip sample on the stm32h573i_dk board, and first it seems to work fine, as it doesn't throw any errors during compilation or running. But there still seems to be a problem (<err> flash_stm32_xspi: XSPI Init failed) with the flash reinitialization during application startup. This problem shows up when adding these lines to prj.conf:
```
CONFIG_FLASH=y
CONFIG_LOG=y
CONFIG_FLASH_LOG_LEVEL_ERR=y
```

But the flash API is used to update and control the images from the application. (Image Manager and similar)

Maybe it would be helpful to test whether the flash API works in the hello_world_xip sample...

Or does anyone have an idea, what needs to be fixed, so the external flash can be used from the application?

etienne-lms commented on drivers/clock_control/clock_stm32_ll_h5.c:
```diff
@@ -435,7 +435,19 @@ static int set_up_plls(void)
 
 #if defined(STM32_PLL_ENABLED)
+
+#if defined(CONFIG_BOOTLOADER_MCUBOOT)
```
I think we rather need a specific config switch (or DT property) to tell Zephur to not touch the PLL.
To be enabled when the Zephyr image executes from an external flash clocked with that PLL.

FRASTM replied:
Don't you think this could be another PR sinc originally this one was for fixing the https://github.com/zephyrproject-rtos/zephyr/issues/87531

FRASTM replied:
see https://github.com/zephyrproject-rtos/zephyr/issues/92879

Jowllel commented:
I just thought that the flash API problem belongs to this PR, because it's used by the mcuboot API and for the image manager. If the flash API doesn't work there's not a direct way to update the mcuboot images. So I think it's the same problem.
Or should I create a new Issue? (But then the Mcuboot only works partially in the zephyr version 4.2.0 when this PR is done.)

erwango commented on samples/boards/st/hello_world_xip/README.rst:
```diff
@@ -0,0 +1,64 @@
+Building and Running
+********************
+
+This application can be built with **sysbuild** and **SB_CONFIG_BOOTLOADER_MCUBOOT**
+West flash will Download the build/mcuboot/zephyr/zephyr.bin at internal flash address 0x08000000
+and the build/hello_world_xip/zephyr/zephyr.signed.bin at internal flash address 0x90000000 (chosen zephyr,code-partition)
```
This is is the same that what is available with `samples/sysbuild/with_mcuboot` I'm not clear on what this new sample is bringing.

FRASTM replied:
yes, more or less
Removing this stm32 sample

FRASTM committed:
soc: stm32: symbol for application XiP from xspi external memory

Add a STM32_APP_IN_EXT_FLASH to determine that an application
is eXecuting in Place on an external xspi flash.
That will control the clock init of this external xspi controller.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
introducing the symbol CONFIG_FLASH_STM32_XSPI_XIP  to identify the application being executed in place on the xspi flash, so that pll clock source of the xspi is not changed when booting that application from mcu boot

FRASTM committed:
drivers: flash: stm32 xspi flash driver skip init when executing in place

The flash_stm32_xspi driver should not initialize the xspi,
if this one is being use to execute in Place : the init is skipped.
This mode is identified with the CONFIG_STM32_APP_IN_EXT_FLASH.
Checking the memory mapped mode bit is possible when the xspi
peripheral clock is not off (stm32h5 has no clock_control_get_status API)

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
rebase on  Zephyr 4.2.0-rc2
change the flash_stm32_xspi driver  to  control the xspi driver init

1.  depending on the `CONFIG_FLASH_STM32_XSPI_XIP` instead of `CONFIG_STM32_MEMMAP`
2.  check the clock_control_get_status: it  returns **CLOCK_CONTROL_STATUS_UNKNOWN** until the clock_stm32_ll_hr driver implements the stm32_clock_control_get_status API function

FRASTM committed:
drivers: clock control for stm32h5 driver implements control_get_status

Add the control_get_status API function
to the stm32h5 clock_control driver

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
Add the get_status API function f to the stm32h5 clock_control driver for the stm32h5 series.

maass-hamburg commented on samples/sysbuild/with_mcuboot/boards/stm32h573i_dk.overlay:
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (c) 2025 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
```
should probably also be added to the hawkbit sample, so the workaround of https://github.com/zephyrproject-rtos/zephyr/pull/92809 can be reversed

maass-hamburg replied:
this overlay should also be added to the hawkbit sample

maass-hamburg commented on boards/st/stm32h573i_dk/stm32h573i_dk.dts:
```diff
@@ -307,21 +307,6 @@
 			reg = <0x00000000 DT_SIZE_K(64)>;
```
maybe a good idea, to increase the mcuboot partition, now that the other space is no longer used

FRASTM replied:
Usually the boot partition is 64K, a next PR could resize but I am not sure this was the purpose here (mainly a bug fix)

FRASTM commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -2056,13 +2059,13 @@ static int flash_stm32_xspi_init(const struct device *dev)
 				     == CLOCK_CONTROL_STATUS_ON) {
 		if (stm32_xspi_is_memorymap(dev)) {
-			LOG_ERR("NOR init'd in MemMapped mode");
+			LOG_INF("NOR init'd in MemMapped mode");
```
changed to LOG_DBG

FRASTM commented on soc/st/stm32/Kconfig.defconfig:
```diff
@@ -85,6 +85,13 @@ config FLASH_BASE_ADDRESS
+# Allows the soc to safely initialize the clocks for the
+# xspi when planning to execute code in xspi Memory.
+# XIP is not set here and depends on each stm32 series (not set for stm32N6)
+config FLASH_STM32_XSPI_XIP
```
Isee, I silly applied the NXP model  but STM32_APP_IN_EXT_MEM sounds good

FRASTM committed:
samples: mgmt: hawkbit: no BOOT_MAX_IMG_SECTORS for stm32h573i_dk

Reverts commit 5700ffe9ad70d4b3d390c3444a39a56db5cfae55.
Not required anymore as the CONFIG_BOOT_MAX_IMG_SECTORS
is fixed by the mcuboot at 4096.
Add the corresponding overlay file.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

etienne-lms commented on soc/st/stm32/Kconfig.defconfig:
```diff
@@ -85,6 +85,14 @@ config FLASH_BASE_ADDRESS
+# Q/O/XSPI controller clocks when the application is residing
+# in external Flash and is chainloaded with MCUboot.
+# Whether the app is eXecuted in Place (XiP) depends on the MCUboot mode used.
+config STM32_APP_IN_EXT_FLASH
```
I think the config should be defined in **Kconfig** with the description in the `help` section.
This **Kconfig.defconfig** file would only set its default value.

etienne-lms commented on drivers/flash/flash_stm32_xspi.c:
```diff
@@ -989,6 +979,19 @@ static int stm32_xspi_abort(const struct device *dev)
+
+	return ((READ_BIT(dev_data->hxspi.Instance->CR,
+			  XSPI_CR_FMODE) == XSPI_CR_FMODE) ?
+			  true : false);
```
By the way:
```suggestion
	return READ_BIT(dev_data->hxspi.Instance->CR, XSPI_CR_FMODE) == XSPI_CR_FMODE;

etienne-lms reviewed (commented):
For consistency, I would suggest to squash 2 commits and reorder some:
1. Commit *"soc: stm32: symbol for application XiP from xspi external memory"*;
2. A single commit that squashes:
  commit *"drivers: clock: stm32H5 clock control of the PLL in MemoryMapped mode"*
  and commit *"drivers: clock control of stm32 devices CONFIG external xspi flash"*;
3. Commit *"drivers: flash: stm32 xspi flash driver skip init when executing in place"*
4. Commit *"boards: arm: stm32h573 disco kit move partition to ext flash"*
5. Commit *"samples: sysbuild: with mcu_boot on the stm32h573i_dk"*
6. Commit *"Revert "samples: mgmt: hawkbit: force BOOT_MAX_IMG_SECTORS for stm32h573i_dk""*

IIUC commit *"drivers: clock control for stm32h5 driver implements control_get_status"* can be placed anywhere in this series.

For commit *"Revert "samples: mgmt: hawkbit: force BOOT_MAX_IMG_SECTORS for stm32h573i_dk""*, I think it would be nice  to add a few words in the commit message to exaplin why it's reverted. By the way, S-o-b tag is missing in that commit message.

FRASTM commented on samples/sysbuild/with_mcuboot/boards/stm32h573i_dk.overlay:
```diff
@@ -0,0 +1,16 @@
+ */
+/ {
+	chosen {
+		zephyr,flash = &mx25lm51245;
```
yes, but I am not sure this is the purpose of this PR fixing the issue for the 4.2.0 release

JarmouniA reviewed (dismissed):
Thank you!

maass-hamburg reviewed (changes requested):
the overlay file is also needed for the hawkbit sample

FRASTM commented:
with the samples/subsys/mgmt/hawkbit/boards/stm32h573i_dk.overlay   included

etienne-lms reviewed (commented):
Non-blocking comments:

1. Nitpicking in commit message for commit *"drivers: clock: stm32H5 clock control of the PLL in MemoryMapped mode"*.
```patch
-Do not disable the pll clock if it source the xspi and if the external
+Do not disable the PLL clock if it sources the XSIP and if the external
 flash is executing in Place. After mcuboot reset, the code is executed
 on the external flash, through the xspi.
-The CONFIG_STM32_APP_IN_EXT_FLASH tis set and will avoid re-config
+The CONFIG_STM32_APP_IN_EXT_FLASH is set and will avoid re-config
-of the pll which is sourcing the xspi peripheral. When excuting in Place
+of the PLL which is sourcing the XSPI peripheral. When executing in Place
-on this external NOR It must not disable its own clock source (PLL).
+on this external NOR, it must not disable its own clock source (PLL).
```
Or simplify:
```
Do not disable the PLL clock if it sources the XSIP and if the external
flash is executing in Place. CONFIG_STM32_APP_IN_EXT_FLASH 
provides this information.
```

2. In commit *"drivers: flash: stm32 xspi flash driver skip init when executing in place"*,
the below sentence in the commit message is no more true since the previous commit implements that API function:
> ... (stm32h5 has no clock_control_get_status API)

3. For commit *"Revert "samples: mgmt: hawkbit: force BOOT_MAX_IMG_SECTORS for stm32h573i_dk""*
This commit does more than reverting a commit. IMHO the commit message header line should not be *"Revert "samples: mgmt: ..."*.

3. Nitpicking on some commit message body: some terminal dots `.` are missing at end of sentences, see commits
*"drivers: flash: stm32 xspi flash driver skip init when executing in place"*,
*"boards: arm: stm32h573 disco kit move partition to ext flash"* and
*"samples: sysbuild: with mcu_boot on the stm32h573i_dk"*.

FRASTM commented on soc/st/stm32/Kconfig:
```diff
@@ -104,4 +104,12 @@ config STM32_BACKUP_PROTECTION
 	  resources needs to be explicitly handled.
 
+config STM32_APP_IN_EXT_FLASH
+	bool
```
no, because of the stm32N6 series

etienne-lms replied:
On N6, if not XIP, the app does not execute in flash so maybe `CONFOIG_STM32_APP_IN_EXT_FLASH` is not a good naming.

etienne-lms replied:
OK thanks.

etienne-lms commented on soc/st/stm32/Kconfig:
```diff
@@ -104,4 +104,12 @@ config STM32_BACKUP_PROTECTION
+	  Allows the SoC clock driver to correctly initialize the
+	  Q/O/XSPI controller clocks when the application is residing
+	  in external Flash and is chainloaded with MCUboot.
+	  Whether the app is eXecuted in Place (XiP) depends on the MCUboot mode used.
```
Ok, let's use `CONFIG_STM32_APP_IN_EXT_FLASH`. Maybe we'll refine the description later.
